odxtools/ecu_variant_matching.py

```diff
@@ -0,0 +1,155 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2023 MBition GmbH
+
+from enum import Enum
+from typing import ByteString, Dict, Generator, List, Optional, Union
+
+from odxtools.diaglayer import DiagLayer
+from odxtools.diaglayertype import DIAG_LAYER_TYPE
+from odxtools.ecu_variant_patterns import MatchingParameter
+from odxtools.service import DiagService
+
+
+class EcuVariantMatcher:
+    """EcuVariantMatcher implements the matching algorithm of ecu variants according to their
+    ECU-VARIANT-PATTERN according to ISO 22901-1.
+
+    Usage (example):
+
+    ```python
+
+    # initialize the matcher with a list of ecu variants,
+    # i.e., DiagLayer instances of variant_type == DIAG_LAYER_TYPE.ECU-VARIANT
+    matcher = EcuVariantMatcher(ecu_variant_candidates=[...], use_cache=use_cache)
+
+    # run the request loop to obtain responses for every request
+    for req in matcher.request_loop():
+        resp = ... # make a UDS request
+        matcher.evaluate(resp)
+
+    # result
+    if matcher.has_match()
+        match = matcher.get_active_ecu_variant()
+    ```
+    """
+
+    class State(Enum):
+        PENDING = 0
+        NO_MATCH = 1
+        MATCH = 2
+
+    @staticmethod
+    def get_ident_service(
+        diag_layer: DiagLayer, matching_param: MatchingParameter
+    ) -> DiagService:
+        service_name = matching_param.diag_comm_snref
+        # TODO this is not working since NamedItemList.__contains__() is not implemented
+        # assert service_name in diag_layer.services
+        service = diag_layer.services[service_name]
+        assert isinstance(service, DiagService)
+        return service
+
+    @staticmethod
+    def encode_ident_request(
+        diag_layer: DiagLayer, matching_param: MatchingParameter
+    ) -> bytes:
+        service = EcuVariantMatcher.get_ident_service(diag_layer, matching_param)
+        return bytes(service.encode_request())
+
+    @staticmethod
+    def decode_ident_response(
+        diag_layer: DiagLayer,
+        matching_param: MatchingParameter,
+        response: Union[bytes, bytearray],
+    ) -> str:
+        service = EcuVariantMatcher.get_ident_service(diag_layer, matching_param)
+        assert service.positive_responses is not None
+        resp_decoded = service.positive_responses[0].decode(response)
+        assert matching_param.out_param_if_snref in resp_decoded
+        return resp_decoded[matching_param.out_param_if_snref]
```

Comment:
Why not just return None if not found?

Reply:
Because I don't want to suggest, that finding the value is optional. The ISO wants the snref or snpathref be resolvable in "at least one" positive response. Let's rather raise an `OdxError` in case no value is found.